### PR TITLE
PyQt5

### PR DESCRIPTION
**Pull request by gamaanderson**
This PR implement support to `PyQt5`, in special as a second try, that is PyQt5 will only be used if importing `PyQt4` fails (but this can be easily changed).

This list of changes between PyQt4 to 5 is present [here](http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html). Since most references to PyQt in the internet are to 4 and not to 5, I do recommend you get familiar with the differences.

The most important difference is that `QtGui` got splitter into `QtWidgets` and `QtGui`, so that must references to `QtGui` are now `QtWidgets`. This got changed all over the code, in PyQt4 mode `QtWidgets` is a copy of `QtGui`.

There were also some differences to the signals and opening files, but this will cause you little to no problem.

### CODE REVIEW

**Comment by nguy** [comment]
I'll start digging into the code and changes soon. My first two questions that come to mind from your statements above:
1. By copying `QtGui` to `QtWidgets` for PyQt4, will there be any conflicts that try to call the classic `QtGui` that are no longer associated?
2. I think that keeping PyQt4 primary and PyQt5 secondary is a good idea for now. Any idea in you looking around the web how much adoption to 5 has taken place?

**Comment by gamaanderson** [comment]
if I have done my work properly in separating what is QtWidgets and what still is QtGui, there should be no conflicts.

Considering that the main doc for PyQt5 is still like (this)[http://pyqt.sourceforge.net/Docs/PyQt5/api/qmainwindow.html], I do not think it is widelly adopted, apple really rushed into that. But maybe the intention was really to bring moviment into the transition (and it worked in our case!).

**Comment by gamaanderson** [comment]
pushed a commit with some changes I forget to commit before

File: artview/core/core.py [hunk 1]
```diff
@@ -7,14 +7,18 @@
...
 
 # lets add some magic for the documentation
-QtCore.__doc__ = ("Qt backend to be used all over ARTview, now it is an "
-                  "alias to :py:mod:`PyQt4.QtCore`")
-QtGui.__doc__ = ("Qt backend to be used all over ARTview, now it is an "
-                 "alias to :py:mod:`PyQt4.QtGui`")
+QtCore.__doc__ = ("Qt backend to be used all over ARTview")
+QtGui.__doc__ = ("Qt backend to be used all over ARTview`")
+QtWidgets.__doc__ = ("Qt backend to be used all over ARTview`, "
+                     "for PyQt4 is a copy of QtGui")
```
Comment (nguy) [inline artview/core/core.py]: I think you may have an extra backtick (`) character in L19 and L20?

**Comment by nguy** [comment]
Still looking through and setting up a special conda environment to install PyQt5 for testing. I'll keep moving forward with this.

**Comment by gamaanderson** [comment]
What I did, it that I installed pyqt5 in python3, so I have both: pyqt4 in python and pyqt5 in python3

**Comment by nguy** [comment]
After setting up my python 3.5 environment the PyQt5 only, the following error on install:
```
/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Traceback (most recent call last):
  File "setup.py", line 219, in <module>
    pre_instalation_tests()
  File "setup.py", line 211, in pre_instalation_tests
    from PyQt4.Qt import PYQT_VERSION_STR
ImportError: No module named 'PyQt4'
```

**Comment by gamaanderson** [comment]
I didnt test the install, I always run local

**Comment by gamaanderson** [comment]
pushed a commit, so setup works, you will still get a lot of warning, but I should setup just fine. I will correct the warnings some other time.

**Comment by nguy** [comment]
Thanks, that did the trick of making the install.

I finally did a clean environment that is running Python 3.5, PyQt5, matplotlib, numpy, scipy, pyart, and artview. I should probably have made a clean test environment like this ages ago for testing.

I'll start looking into the code a little deeper now.

**Comment by nguy** [comment]
I don't know if this is a function of this PR, but I am getting the following error during testing on the Python3/PyQt5 side. Can you take a look?

```  
File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/plot_radar.py", line 347, in <lambda>
    lambda RngRing=RngRing: self.RngRingSelectCmd(RngRing))
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/plot_radar.py", line 646, in RngRingSelectCmd
    self.RNG_RINGS = range(ringdel, unrng, ringdel)
UnboundLocalError: local variable 'ringdel' referenced before assignment
```

**Comment by nguy** [comment]
Also saw this problem:
```
Traceback (most recent call last):
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/plot_radar.py", line 553, in NewCmap
    self._update_plot()
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/plot_radar.py", line 933, in _update_plot
    ax=self.ax, fig=self.fig, title=title)
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/pyart/graph/radardisplay.py", line 404, in plot_ppi
    x, y, data, vmin=vmin, vmax=vmax, cmap=cmap, norm=norm, **kwargs)
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/__init__.py", line 1819, in inner
    return func(ax, *args, **kwargs)
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/axes/_axes.py", line 5395, in pcolormesh
    collection.set_cmap(cmap)
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/cm.py", line 306, in set_cmap
    cmap = get_cmap(cmap)
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/matplotlib/cm.py", line 166, in get_cmap
    % (name, ', '.join(sorted(cmap_d.keys()))))
ValueError: Colormap False is not recognized. Possible values are: Accent, Accent_r, Blues, Blues_r, BrBG, BrBG_r, BuGn, BuGn_r, BuPu, BuPu_r, CMRmap, CMRmap_r, Dark2, Dark2_r, GnBu, GnBu_r, Greens, Greens_r, Greys, Greys_r, OrRd, OrRd_r, Oranges, Oranges_r, PRGn, PRGn_r, Paired, Paired_r, Pastel1, Pastel1_r, Pastel2, Pastel2_r, PiYG, PiYG_r, PuBu, PuBuGn, PuBuGn_r, PuBu_r, PuOr, PuOr_r, PuRd, PuRd_r, Purples, Purples_r, RdBu, RdBu_r, RdGy, RdGy_r, RdPu, RdPu_r, RdYlBu, RdYlBu_r, RdYlGn, RdYlGn_r, Reds, Reds_r, Set1, Set1_r, Set2, Set2_r, Set3, Set3_r, Spectral, Spectral_r, Wistia, Wistia_r, YlGn, YlGnBu, YlGnBu_r, YlGn_r, YlOrBr, YlOrBr_r, YlOrRd, YlOrRd_r, afmhot, afmhot_r, autumn, autumn_r, binary, binary_r, bone, bone_r, brg, brg_r, bwr, bwr_r, cool, cool_r, coolwarm, coolwarm_r, copper, copper_r, cubehelix, cubehelix_r, flag, flag_r, gist_earth, gist_earth_r, gist_gray, gist_gray_r, gist_heat, gist_heat_r, gist_ncar, gist_ncar_r, gist_rainbow, gist_rainbow_r, gist_stern, gist_stern_r, gist_yarg, gist_yarg_r, gnuplot, gnuplot2, gnuplot2_r, gnuplot_r, gray, gray_r, hot, hot_r, hsv, hsv_r, inferno, inferno_r, jet, jet_r, magma, magma_r, nipy_spectral, nipy_spectral_r, ocean, ocean_r, pink, pink_r, plasma, plasma_r, prism, prism_r, pyart_BlueBrown10, pyart_BlueBrown10_r, pyart_BlueBrown11, pyart_BlueBrown11_r, pyart_BrBu10, pyart_BrBu10_r, pyart_BrBu12, pyart_BrBu12_r, pyart_Bu10, pyart_Bu10_r, pyart_Bu7, pyart_Bu7_r, pyart_BuDOr12, pyart_BuDOr12_r, pyart_BuDOr18, pyart_BuDOr18_r, pyart_BuDRd12, pyart_BuDRd12_r, pyart_BuDRd18, pyart_BuDRd18_r, pyart_BuGr14, pyart_BuGr14_r, pyart_BuGy8, pyart_BuGy8_r, pyart_BuOr10, pyart_BuOr10_r, pyart_BuOr12, pyart_BuOr12_r, pyart_BuOr8, pyart_BuOr8_r, pyart_BuOrR14, pyart_BuOrR14_r, pyart_Carbone11, pyart_Carbone11_r, pyart_Carbone17, pyart_Carbone17_r, pyart_Carbone42, pyart_Carbone42_r, pyart_Cat12, pyart_Cat12_r, pyart_EWilson17, pyart_EWilson17_r, pyart_GrMg16, pyart_GrMg16_r, pyart_Gray5, pyart_Gray5_r, pyart_Gray9, pyart_Gray9_r, pyart_LangRainbow12, pyart_LangRainbow12_r, pyart_NWSRef, pyart_NWSRef_r, pyart_NWSVel, pyart_NWSVel_r, pyart_NWS_SPW, pyart_NWS_SPW_r, pyart_PD17, pyart_PD17_r, pyart_RRate11, pyart_RRate11_r, pyart_RdYlBu11b, pyart_RdYlBu11b_r, pyart_RefDiff, pyart_RefDiff_r, pyart_SCook18, pyart_SCook18_r, pyart_StepSeq25, pyart_StepSeq25_r, pyart_SymGray12, pyart_SymGray12_r, pyart_Theodore16, pyart_Theodore16_r, pyart_Wild25, pyart_Wild25_r, rainbow, rainbow_r, seismic, seismic_r, spectral, spectral_r, spring, spring_r, summer, summer_r, terrain, terrain_r, viridis, viridis_r, winter, winter_r
```

File: artview/components/image_text.py [hunk 2]
```diff
@@ -267,16 +267,12 @@ def _init_entries(self):
...
-        if not isinstance(str(self.ent_texc.text()), str):
-            core.common.ShowWarning('Color must be a valid string ')
-        if not isinstance(int(self.ent_texsz.text()), int):
-            core.common.ShowWarning('Text size must be integer value')
+        try: float(self.ent_xpos.text())
+        except ValueError: core.common.ShowWarning('X-Coord must be float value')
+        try: float(self.ent_ypos.text())
+        except ValueError: core.common.ShowWarning('Y-Coord must be float value')
+        try: int(self.ent_texsz.text())
+        except ValueError: core.common.ShowWarning('Text size must be integer value')
```
Comment (nguy) [inline artview/components/image_text.py]: This looks a lot cleaner.

**Comment by nguy** [comment]
I'm experiencing a lot of crashes with this new code. For example just testing the new error catching in `image_text`, the program crashes.
```
Traceback (most recent call last):
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/image_text.py", line 188, in updateChoice
    self.choice = self._get_entries()
  File "/Users/guy/anaconda/envs/py35_env/lib/python3.5/site-packages/artview/components/image_text.py", line 283, in _get_entries
    'xpos': float(self.ent_xpos.text()),
ValueError: could not convert string to float: 'big'
Abort trap: 6
```

File: artview/components/menu.py [hunk 3]
```diff
@@ -159,9 +159,12 @@ def removeTab(self, idx):
...
         '''Open a dialog box to choose file.'''
 
-        filename = QtGui.QFileDialog.getOpenFileName(
+        filename = QtWidgets.QFileDialog.getOpenFileName(
             self, 'Open file', self.dirIn)
+        if isinstance(filename, tuple): # PyQt5
+            filename = filename[0]
         filename = str(filename)
+        print("filename",filename.__repr__())
         if filename == '':
```
Comment (nguy) [inline artview/components/menu.py]: Is this print statement left over from testing?

File: artview/components/plot_grid.py [hunk 4]
```diff
@@ -8,15 +8,19 @@
...
-from matplotlib.backends.backend_qt4agg import NavigationToolbar2QT as \
-    NavigationToolbar
+from matplotlib.backends import pylab_setup
+backend = pylab_setup()[0]
+FigureCanvasQTAgg = backend.FigureCanvasQTAgg
+NavigationToolbar = backend.NavigationToolbar2QT
+#from matplotlib.backends.backend_qt4agg import FigureCanvasQTAgg
+#from matplotlib.backends.backend_qt4agg import NavigationToolbar2QT as \
+#    NavigationToolbar
 from matplotlib.figure import Figure
```
Comment (nguy) [inline artview/components/plot_grid.py]: Is this an updated method to call these classes?

**Comment by nguy** [comment]
Also experiencing trouble selecting the `File` and `Modes` menus.

File: artview/components/menu.py (hunk 3 above)
Comment (gamaanderson) [inline artview/components/menu.py]: no, this is just garbage

File: artview/components/plot_grid.py (hunk 4 above)
Comment (gamaanderson) [inline artview/components/plot_grid.py]: matplotlib do not explain how to access the backend. I found by looking in the code this should be the most high level call, and therefore backend independent. Matplotlib is just a mess!

**Comment by gamaanderson** [comment]
I have pushed a new commit correct the points you present, some of those were also in the present before. I could however no reproduce the `Abort trap: 6` crash. It [seems like](http://stackoverflow.com/questions/29309016/bundling-a-pyqt5-application-using-py2app-keep-getting-abort-trap-6-error) this is a bug of the OSX. In the end this is a C allocation error, python should not crash in this way.

**Comment by nguy** [comment]
I pulled the updates and will test some more.
Which errors were present before? Just trying to keep track.

**Comment by gamaanderson** [comment]
there where an error in connecting the action for drawing circle and change colormap. There as also the error that _check_entries did not interrupt further execution, so _get_entries would fail.

**Comment by nguy** [comment]
Thanks, I remember putting the entries on my to-do list, but either had not noticed or forgotten the others.

If the MacOSX bug continues to be a problem, I'm not sure what to do. I suspect that many users work on Macbooks, though I don't know that for certain. 
I will do more testing and try the following as suggested in the link you sent `python3 setup.py py2app -A --packages=PyQt5` if I continue to have problems. If it solves the issue, then we may need to get creative with the install script.

**Comment by nguy** [comment]
And just to clarify, you don't see this issue at all on your linux machine, correct?

**Comment by gamaanderson** [comment]
yes, I got the python error (that is know corrected), but not the subsequent crash. Once it is running, I hardly ever got a crash of artview.

**Comment by nguy** [comment]
Sorry for the delay, I've been fussing with this on and off for the last month with no luck in getting rid of the error. The macdeployqt solution suggested in a link above seems to be a solution for packing the application when it gets working. I've tried compiling with different flags and such. No luck.

The "Abort Trap 6" error happens when trying to use many plugins and tools. I need to figure a more robust way to test and to trace the problem for more information. I'll be back to working on this more.

Also a weird one: The File menu is not in the window as it was previously. And on top of this it is not selectable at startup. I have to click on the Terminal window where I started ARTview and then I can access the menu. Odd behavior.

**Comment by gamaanderson** [comment]
I would say we put an comment in the README about this problem, but do not allow it to delay the PR. With some luck this will get solved in the future. Could you prepare such comment.

**Comment by nguy** [comment]
I will. 

Do you think we should include a way to use PyQt4 instead of 5 or put that information in the README for reference as well?

**Comment by gamaanderson** [comment]
pyqt4 will always be preferred over pyqt5, that is already the way it is.

**Comment by gamaanderson** [comment]
just pushed the new readme

**Comment by nguy** [comment]
Nice, merging to get this through.
Just for the record, this does only occur with PyQt5 on a MacOS. Python 2/3 don't matter, along with any combination with Qt4 works fine.

**Pull request merged**